Improve error stack traces for async errors

rahulchavan30 opened the pull request:
<!--
Thanks for your interest in the project. Bugs filed and PRs submitted are appreciated!

Please make sure that you are familiar with and follow the Code of Conduct for
this project (found in the CODE_OF_CONDUCT.md file).

Also, please make sure you're familiar with and follow the instructions in the
contributing guidelines (found in the CONTRIBUTING.md file).

If you're new to contributing to open source projects, you might find this free
video course helpful: https://kcd.im/pull-request

Please fill out the information below to expedite the review and (hopefully)
merge of your pull request!
-->

<!-- What changes are being made? (What feature/bug is being fixed here?) -->

**What**:
Improve error stack traces for async errors
<!-- Why are these changes necessary? -->

**Why**:
Improve error stack traces for async errors
<!-- How were these changes implemented? -->

**How**:
-changed the on timeout function  
<!-- Have you done all of these things?  -->

**Checklist**:

<!-- add "N/A" to the end of each line that's irrelevant to your changes -->

<!-- to check an item, place an "x" in the box like so: "- [x] Documentation" -->

- [ ] Documentation added to the
      [docs site](https://github.com/testing-library/testing-library-docs)
- [ ] I've prepared a PR for types targeting
      [DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/testing-library__dom)
- [ ] Tests
- [ ] Ready to be merged
      <!-- In your opinion, is this ready to be merged as soon as it's reviewed? -->

<!-- feel free to add additional comments -->
i have made changes suggested by KCD , please let me know if i need to do any more changes will be happy to do ... this is my first PR to Open source

marcosvega91 commented:
Hi @rahulchavan30 good job for your first contribution 🥳 . As I understand this behavior should be enabled based on a config parameter.
I think that you can add a new property in src/config.js and use it in wait-for.js.
Maybe It's good to add also a test case.  
Last thing, to make the PR visibile in the issue you can add a reference like #539.

Regards
Marco

rahulchavan30 commented:
hi macro , 
Thanks for the update ... if i understand clearly i need to do the following  ,

1. add a config parameter showOriginalStackTrace .
2. by default showOriginalStackTrace will be false and when its false i need to return 
onDone(lastError || timedOutError, null)  if its true i need to return  
let error = timedOutError
  if (lastError) {
    error = lastError
    const userStackTrace = timedOutError.stack.split('\n').slice(1).join('\n')
    error.stack = `${error.stack.split('\n')[0]}\n${userStackTrace}`
  }
  onDone(error, null); 
i will mae these  changes and update the PR if you agree ..

marcosvega91 commented:
Hi @rahulchavan30 You're welcome 😸 
Yes that's right. 
You can put only this 

```
const userStackTrace = timedOutError.stack.split('\n').slice(1).join('\n')
error.stack = ${error.stack.split('\n')[0]}\n${userStackTrace}
```

Around  the new if. I think it's more readable

rahulchavan30 commented:
Thanks macro will do the changes and submit ...

rahulchavan30 commented:
Hi Macro , please bear with me , i have made the necessary changes but when i run npm test update  
i get an error 

i have changed the code to the following 

function onTimeout() {
      let error = timedOutError
      let showOriginalStackTrace = getConfig().showOriginalStackTrace
      if (lastError) {
        error = lastError
        if(showOriginalStackTrace){
          const userStackTrace = timedOutError.stack.split('\n').slice(1).join('\n')
          error.stack = `${error.stack.split('\n')[0]}\n${userStackTrace}`
        }
      }
      onDone(error, null)
    }

but i get this code coverage error  wait-for.js                       |   94.74 |    94.12 |     100 |   94.29 | 71-72 
can you help me out with the test i need to satisfy the code coverage ?

rahulchavan30 commented:
Quick update i added a new test case and was able to get the code coverage passing

kentcdodds commented on src/__tests__/wait-for.js:
```diff
@@ -34,6 +34,17 @@ test('uses generic error if there was no last error', async () => {
+    },
+    {timeout: 8, interval: 5, showOriginalStackTrace: true},
+  ).catch(e => e)
+  expect(result).toBe(error)
```
It's possible that the `error.stack` could've been manipulated (if we decide to change the implementation in the future). So I'd change this:

```suggestion
  expect(result.stack).toBe(error.stack)
```

kentcdodds commented on src/__tests__/wait-for.js:
```diff
@@ -34,6 +34,17 @@ test('uses generic error if there was no last error', async () => {
+  ).catch(e => e)
+  expect(result.stack).toBe(error.stack)
+})
+
```
Could we add one test to verify that the stack trace is set correctly?

This might be tricky though. If you struggle with it, go ahead and push what you have and we can help :)

eps1lon replied:
This is especially interesting for different node versions in case node changes how they display stack traces.

rahulchavan30 replied:
i am not knowing how to proceed .. any direction of help will be appreciated .. from what i understood you want me to add a new test that verifies the stack ?

eps1lon replied:
I think a test that shows the actual trace would be nice. You'd probably have to strip the parts of the filepaths that are tied to your local filesystem. You also want to only match the last significant stacks. Otherwise we're probably matching node internals.

rahulchavan30 replied:
i am sorry i tried alot but i am not sure how to proceed .

kentcdodds replied:
No worries @rahulchavan30! We can help with this part. Thanks for all you've done :)

rahulchavan30 replied:
Thank you Kent and thank you so much for the opportunity I really got to learn so many things in short time of span

kentcdodds replied:
That's wonderful to hear :) Take a look at my changes. It ended up being a bit challenging for me as well, but I'm happy with what we're going to ship together :)

kentcdodds commented:
@all-contributors please add @rahulchavan30 for code and tests

kentcdodds commented:
:tada: This PR is included in version 7.3.0 :tada:

The release is available on:
- `npm package (@latest dist-tag)`
- [GitHub release](https://github.com/testing-library/dom-testing-library/releases/tag/v7.3.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
